orc/src/main/java/org/apache/iceberg/orc/OrcSchemaWithTypeVisitor.java

```diff
@@ -78,14 +79,73 @@ protected T visitUnion(Type type, TypeDescription union, OrcSchemaWithTypeVisito
+  as the corresponding types in the union of Orc schema.
+  Except the tag field, the fields in the struct of Iceberg schema are the same as the types in the union of Orc schema
+  in the general case. In case of field projection, the fields in the struct of Iceberg schema only contains
+  the fields to be projected which equals to a subset of the types in the union of ORC schema.
+  Therefore, this function visits the complex union with the consideration of both cases.
+   */
+  private <T> void visitComplexUnion(Type type, TypeDescription union, OrcSchemaWithTypeVisitor<T> visitor,
+                                     List<T> options) {
+    int typeIndex = 0;
+    int fieldIndexInStruct = 0;
+    while (typeIndex < union.getChildren().size()) {
+      TypeDescription schema = union.getChildren().get(typeIndex);
+      boolean relatedFieldInStructFound = false;
+      Types.StructType struct = type.asStructType();
+      if (fieldIndexInStruct < struct.fields().size() &&
+              ORCSchemaUtil.ICEBERG_UNION_TAG_FIELD_NAME
+                      .equals(struct.fields().get(fieldIndexInStruct).name())) {
+        fieldIndexInStruct++;
+      }
+
+      if (fieldIndexInStruct < struct.fields().size()) {
+        String structFieldName = type.asStructType().fields().get(fieldIndexInStruct).name();
+        int indexFromStructFieldName = Integer.parseInt(structFieldName.substring(5));
+        if (typeIndex == indexFromStructFieldName) {
+          relatedFieldInStructFound = true;
+          T option = visit(type.asStructType().fields().get(fieldIndexInStruct).type(), schema, visitor);
+          options.add(option);
+          fieldIndexInStruct++;
+        }
+      }
+      if (!relatedFieldInStructFound) {
+        visitNotProjectedTypeInComplexUnion(schema, visitor, options, typeIndex);
+      }
+      typeIndex++;
+    }
+  }
+
+  // If a field is not projected, a corresponding field in the struct of Iceberg schema cannot be found
+  // for current type of union in Orc schema, a reader for current type still needs to be created and
+  // used to make the reading of Orc file successfully. In this case, a pseudo Iceberg type is converted from
+  // the Orc schema and is used to create the option for the reader of the current type which still can
+  // read the corresponding content in Orc file successfully.
+  private static <T> void visitNotProjectedTypeInComplexUnion(TypeDescription schema,
+                                                              OrcSchemaWithTypeVisitor<T> visitor,
+                                                              List<T> options,
+                                                              int typeIndex) {
+    OrcToIcebergVisitor schemaConverter = new OrcToIcebergVisitor();
+    schemaConverter.beforeField("field" + typeIndex, schema);
+    schema.setAttribute(org.apache.iceberg.orc.ORCSchemaUtil.ICEBERG_ID_ATTRIBUTE, "-1");
```

Comment:
this -1 can also be created as an 'invalid field id" constant. That would also clarify the objective - multiple fields can have an invalid field id.

Reply:
Good point, done